composer/loggers/tensorboard_logger.py

```diff
@@ -75,20 +79,34 @@ def log_data(self, state: State, log_level: LogLevel, data: Dict[str, Any]):
                 pass
 
     def init(self, state: State, logger: Logger) -> None:
-        from torch.utils.tensorboard import SummaryWriter
+        self.run_name = state.run_name
 
         # We set the log_dir to a constant, so all runs can be co-located together.
         if self.log_dir is None:
             self.log_dir = 'tensorboard_logs'
 
+        self._initialize_summary_writer()
+
+    def _initialize_summary_writer(self):
+        from torch.utils.tensorboard import SummaryWriter
+
+        assert self.run_name is not None
+        assert self.log_dir is not None
         # We name the child directory after the run_name to ensure the run_name shows up
         # in the Tensorboard GUI.
-        summary_writer_log_dir = Path(self.log_dir) / state.run_name
+        summary_writer_log_dir = Path(self.log_dir) / self.run_name
 
         # To disable automatic flushing we set flushing to once a year ;)
         flush_secs = 365 * 3600 * 24
         self.writer = SummaryWriter(log_dir=summary_writer_log_dir, flush_secs=flush_secs)
```

Comment:
why can't we just use SummaryWriter's own flushing frequency? Can we add a concise comment here?

Reply:
The tl;dr is the summarywriter flush happens in a background thread, and that could corrupt the tracefile if the background thread is flushing at the same time that the object store logger is staging the trace file for upload. Agree this should be elaborated on in a comment.

Reply:
done